### PR TITLE
Stress test

### PR DESCRIPTION
**Pull request by AYAHASSAN287**
**Summary**

This PR adds several long-running endurance tests to exercise the Waku nodes under various scenarios. We also introduce three separate GitHub Actions workflows (in three .yml files) for different CI jobs:

**Endurance Test 1**

Runs all tests in a loop (up to 12 hours) to observe any potential resource leaks or performance degradation.

**Endurance Test 2**

Runs specific endurance scenarios (e.g., high-churn, large payload ephemeral messages, PX load tests) for extended periods.

**Other Tests**

Runs the remaining standard tests (non-endurance) for quicker feedback on code changes.

**Details**

The new endurance tests focus on memory monitoring but currently do not enforce pass/fail based on memory usage.

Assertions on memory usage are disabled to avoid failing the tests due to known issues. A separate maintenance task will be created to refine these assertions and address known memory edge cases.

**Note**

-Assertions disabled as many tests failing currently for >1k iteration so follow up task for maintanance will be opened to address the issues of the tests 

- to run the 12h job a self hosted runner or more powerful machine will be used so this point is not fully done yet and will be finished in a seperate task

### CODE REVIEW

**Comment by gabrielmer** [comment]
Now that I think about it, maybe we shouldn't have the stress tests in the CI that runs in every PR? Because it would make the CI runs way too long and we would have to wait for the following day to merge a PR. Maybe just having them run as a nightly job?

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+name: Endurance Tests
+
+on:
+  push:
+    branches: [ "stress_test" ]
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply .github/workflows/CI.yml:5]: we can have it now for the normal job but the 2 endurance job aren't currently successfully running till end so we can add it later

File: waku/test_repeated_start_stop.log [hunk 2]
```diff
@@ -0,0 +1,62 @@
+time="2025-03-11T11:42:39+02:00" level=debug msg="[Test5Nodes1kTearDown] Memory usage at test START: 826 KB"
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/test_repeated_start_stop.log:1]: will ignore it

File: waku/endurance_readings.log
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/endurance_readings.log:1]: ignored

File: waku/peer_connections_test.go [hunk 3]
```diff
@@ -134,6 +134,7 @@ func TestDiscv5PeerMeshCount(t *testing.T) {
 	Debug("Waiting for nodes to auto-connect via Discv5")
 	err = WaitForAutoConnection([]*WakuNode{node1, node2, node3})
 	require.NoError(t, err, "Nodes did not auto-connect within timeout")
+	time.Sleep(time.Second * 10)
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/peer_connections_test.go:137]: the delay was  less than 10 and worked fine locally but on CI it fails many times . when increased delay to 10 it worked on CI as well  but we can reduced it to 5 and add the issue to the follow up task if still happens

File: waku/stress_test.go [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+
+	time.Sleep(2 * time.Second)
+
+	node1.StopAndDestroy()
+	node2.StopAndDestroy()
+	node3.StopAndDestroy()
+
+	runtime.GC()
+	time.Sleep(1 * time.Second)
+	runtime.GC()
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/stress_test.go:65]: the reason is that finalizers, goroutines, or other asynchronous cleanup tasks may be triggered by the first garbage-collection cycle but not necessarily complete fast enough so adding it again after sleep time just to ensure all garbage collected

File: waku/stress_test.go [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+	runtime.ReadMemStats(&memStats)
+	Debug("[%s] Memory usage AFTER destroying nodes: %d KB", testName, memStats.HeapAlloc/1024)
+
+	Debug("[%s] Test completed successfully", testName)
+}
+
+func TestStressStoreQuery5kMessagesWithPagination(t *testing.T) {
+	Debug("Starting test")
+	runtime.GC()
+	time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/stress_test.go:76]: I think I added the wait here because I was debugging the test it was failing and I was looking for possible fixes but now won't need it so I will delete it

**Comment by AYAHASSAN287** [comment]
> Now that I think about it, maybe we shouldn't have the stress tests in the CI that runs in every PR? Because it would make the CI runs way too long and we would have to wait for the following day to merge a PR. Maybe just having them run as a nightly job?

agreed I change it to run when triggered manually . it shall be part of the nightly but when we have the 2 jobs run successfully as now we have some blockers due to machine limitations and tests instability

**Review by gabrielmer — APPROVED** [review]
Thanks so much!

File: waku/relay_test.go [hunk 6]
```diff
@@ -292,12 +292,13 @@ func TestRelaySubscribeAndPeerCountChange(t *testing.T) {
...
+	// Commented till we configure external IPs
+	//node2Address, err := node2.ListenAddresses()
+	//require.NoError(t, err, "Failed to get listening address for Node2")
 
 	node3Config := DefaultWakuConfig
 	node3Config.Relay = true
-	node3Config.Staticnodes = []string{node2Address[0].String()}
+	node3Config.Staticnodes = []string{node1Address[0].String()}
 
 	Debug("Creating Node3 with Node2 as a static node")
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/relay_test.go]: done

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+name: Endurance Tests
```
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply .github/workflows/CI.yml]: NO, Changed

File: .github/workflows/CI.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply .github/workflows/CI.yml]: Done for workflow dispatch and cron at 2 am

File: waku/test_repeated_start_stop.log (hunk 2 above)
[the remark this answers is not in the document]
Comment (AYAHASSAN287) [reply waku/test_repeated_start_stop.log:1]: done
